spyql/processor.py

```diff
@@ -457,16 +466,49 @@ def get_input_iterator(self):
 
 # CSV
 class CSVProcessor(Processor):
-    def __init__(self, prs, strings, sample_size=10, header=None, **options):
+    def __init__(
+        self, prs, strings, sample_size=10, header=None, infer_dtypes=True, **options
+    ):
         super().__init__(prs, strings)
         self.sample_size = sample_size
         self.has_header = header
+        self.infer_dtypes = infer_dtypes
         self.options = options
         csv.reader(StringIO("test"), **self.options)  # test options
 
+    def _test_dtype(self, v):
+        v = v.strip()
+        if not v:
+            return (-100, None)  # empty string: do not cast
+        try:
+            int(v)
+            return (10, "int")
+        except ValueError:
+            try:
+                float(v)
+                return (20, "float")
+            except ValueError:
+                try:
+                    complex(v)
+                    return (30, "complex")
+                except ValueError:
+                    return (100, None)  # not a basic type: do not cast
+
+    def _infer_dtypes(self, reader):
+        if self.has_header:
+            next(reader, None)  # skip header
+        dtypes_rows = [[self._test_dtype(col) for col in line] for line in reader]
+        if dtypes_rows and dtypes_rows[0]:
+            dtypes = [
+                max([row[c] for row in dtypes_rows]) for c in range(len(dtypes_rows[0]))
+            ]
```

Comment:
I'm a little worried about the impact this may have on the performance. I understand that to have the most accurate type inference we need as much data as possible but is it worth the performance penalty?

Reply:
When reading CSVs, we start by reading a sample (10 lines by default) into memory. It is on this sample that we detect if there's a header and detect the dialect (e.g. the delimiter). It is in this sample that we run the type inference:
https://github.com/dcmoura/spyql/blob/436e3e7108ada5419cab6bcda612434141553c0f/spyql/processor.py#L538

Since this sample is typically small (less than 100 lines), I don't expect a big performance penalty. Do you agree?